main/src/library/Prelude.flix

```diff
@@ -164,3 +164,23 @@ pub def printUnlessUnit(x: a): Unit \ IO with ToString[a] = {
         case _: _ => println(x)
     }
 }
+
+pub enum Purity[a: Type, b: Type, ef: Eff] {
+    case Pure(a -> b)
+    case Impure(a -> b \ ef)
+}
+
+pub def purityOf(f: a -> b \ ef): Purity[a, b, ef] = typematch f {
+    case g: a -> b \ Pure => Purity.Pure(g)
+    case _: _             => Purity.Impure(f)
+}
+
+pub def purityOf2(f: a -> b -> c \ ef): Purity[(a, b), c, ef] = typematch f {
+    case g: (a, b) -> c \ Pure => Purity.Pure(match (x, y) -> g(x, y))
+    case _: _                  => Purity.Impure(match (x, y) -> f(x, y))
+}
+
+pub def purityOf3(f: a -> b -> c -> d \ ef): Purity[(a, b, c), d, ef] = typematch f {
+    case g: (a, b, c) -> d \ Pure => Purity.Pure(match (x, y, z) -> g(x, y, z))
+    case _: _                  => Purity.Impure(match (x, y, z) -> f(x, y, z))
+}
```

Comment:
These need documentation.

Could we do the currying thing inside these functions so that the user of the functions doesn't have to handle any currying stuff?

Reply:
Ah yes. Brilliant idea!

Reply:
> These need documentation.
> 
> Could we do the currying thing inside these functions so that the user of the functions doesn't have to handle any currying stuff?

I tried experimenting with currying inside `purityOf` but I couldn't make it work. It infers `a -> (b -> c) \ ef` where `a -> (b -c \ ef) is expected`